File: DailyTemp.py
#Time Complexity : O(n)
#Space Complexity :O(n)
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        if temperatures == None or len(temperatures) == 0:
            return 0
        st = []
        result = [0] * len(temperatures)
        for i in range(len(temperatures)):
            while len(st) != 0 and temperatures[st[-1]] < temperatures[i]:
                x = st.pop()
                result[x] = i - x
            st.append(i)
        return result

File: daily_temp.java
//Optimal : Using Stack

class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int[] res = new int[temperatures.length];
        
        if(temperatures == null || temperatures.length == 0)
            return res;
        
        Stack<Integer> st = new Stack<>();
        
        for(int i=0; i<temperatures.length; i++){
            while(!st.isEmpty() && temperatures[i] > temperatures[st.peek()]){
                int diff = i - st.peek();
                res[st.peek()] = diff;
                st.pop();
            }
            st.push(i);
        }
        return res;
    }
}

//Time Complexity :O(n)
//Space Complexity :O(n)